### DISCUSSION TITLE
What's the goal of Nu as a language and package manager

### DISCUSSION THREAD

**Discussion by nxy7** [nushell/nushell#9504] [General]
Hello, I'm using Nushell as my main shell for quiet a while and I've some questions that I've found no answers to.

It's probably not a stretch to say that most people see Nushell mainly as a shell, this discussion however is more about Nu language. 
At first I thought its mostly made as a tool to write scripts, but there's plenty of other utilities here that make me think that the end goal here might actually be some sort of Python/NodeJS replacement. Having package manager in development suggests the same, after all why create package manager for simple scripts - I don't think that would make much sense. 

If that's actually a case I think killer feature (something that makes it different enough that people would use Nushell just for that one goal) of Nu could be making CLI apps. It's something I and some other people have suggested and imao Nu is uniquely positioned here to be the best tool for the job. It already has great self documenting export system, treats executables as first class objects, has fast iteration speed (as it's interpreted), all that's left is somehow compiling said scripts to one binary (i don't even think that shipping whole nushell would be bad, but I've read one comment about translating nu files to rust which would also give it unmatched performance). I can't think of anything that would be even close to Nu in that space. As CLI apps could get quiet complex it makes sense to have package manager to share some code with other people easily.

I'm not sure if there are other valid use cases right now, as ecosystem is still small and young why would someone pick nu over python with gazillion libraries ready to use. I'm really curious how and why do you think people will consume nu packages? (hope it doesn't sound as harsh question as I really love Nushell and it's genuine question)

I'm not nearly smart enough to contribute here so I've opened this discussion just to know what core maintainers think about this.

TL;DR: What's Nu language end goal and what use cases do you see for package manager?

**Comment by fdncred** [comment]
I can only speak for myself, so here are some thoughts I have.

> nupm

If you've ever looked at our nu_scripts repo, you'll see many scripts, modules, themes, completions, etc. We've always thought it would be nice to have a well-rounded experience in nushell where people could just say, "nushell, please show me themes", or "nushell, please install the dracula theme". (not necessarily using natural language). In order to achieve such a goal, you have to have the building blocks, which we pretty much have now. You have to have some way to reach the web, we have `http get`. You have to have some way to describe and deserialize information, we support nuon/json/yaml/toml/csv/etc. The piece that is still missing is gluing all this together, and, in my mind, that's kind of what nupm is.

> standalone cli apps

Agreed, that would be an awesome feature. We've talked about it several times and created a prototype. I think the one missing thing right now is stability. IMO, I think we need a stable API and platform before we get too far down the road of how to do this type of thing. I'm hopeful that once we hit 1.0, these types of things will be stable, and it will be easier to do. 

I see a couple paths to this, once we have the stability required. (1) Just use the nushell crates, analyze and add a script to see what commands it's using, then create a default context that has those commands, and use rust to compile an executable that embeds that script and runs it. (2) I think this is more complicated, but generating bytecode and assembling it is the other option that jumps out at me.

> goals

My overarching goal is to help create an awesome cross platform tool, that uses structured data, and that people love and use daily. I don't plan to take over the world, but I do plan to make lives easier when using the console.

**Reply by nxy7** [reply]
Okay, thanks for clearing things up a bit. So if I understand correctly nupm is not totally analogous to npm as it will be able to integrate into shell, so you could use it to extend your shell capabilities, not just use it with your script files. That makes much more sense and sounds pretty cool.

Thank you for taking time to respond to my question.

**Reply by amtoine** [reply]
> Okay, thanks for clearing things up a bit.
>
i agree, @fdncred's answer is very good and i can only support what has been said :pray: 

> So if I understand correctly nupm is not totally analogous to npm as it will be able to integrate into shell, so you could use it to extend your shell capabilities, not just use it with your script files. That makes much more sense and sounds pretty cool.
> 
yes this is one of the intent of `nupm` :+1: 
at some point maybe it will allow for a *real* programming experience like with `npm`, `node`, pip` or `cargo`? we'll see in the future i guess :relieved: 

we have been at least three, myself included, to write a prototype version of a *Nushell package manager*, so we decided to merge the ideas into https://github.com/nushell/nupm.
when i wrote mine, it was mainly to install scripts of mine that were in other repos, e.g. https://github.com/goatfiles/scripts/tree/main/nu_scripts or https://github.com/amtoine/nu-git-manager.
the goal being mainly to avoid doing the whole install and updates manually :smirk: 
same goas with the themes and completions from Nushell in `nushell/nu_scripts` which can be in theory installed in the same way :ok_hand:

**Comment by balupton** [comment]
I'm interested in following along. I've authored hundreds of bash commands as part of https://github.com/bevry/dorothy and considering adding nushell to its shell support, and perhaps even writing commands within it using nushell.

One of the limitations with its bash commands approach, is that they aren't that portable. It would be so cool to be able to compile a .nush script and deploy it anywhere.
